src/Streamly/Internal/Data/Fold/Types.hs

```diff
@@ -652,19 +652,67 @@ longest :: -- Monad m =>
     Fold m x a -> Fold m x a -> Fold m x a
 longest _f1 _f2 = undefined
 
+data ConcatMapState m sa a c
+    = B !sa
+    | forall s. C (s -> a -> m (Step s c)) !s (s -> m c)
+
 -- | Map a 'Fold' returning function on the result of a 'Fold'.
 --
--- Compare with 'Monad' instance method '>>='. This implementation allows
--- stream fusion but has quadratic complexity. This can fuse with other
--- operations and can be much faster than 'Monad' instance for small number
--- (less than 8) of compositions.
+-- Compare with 'Monad' instance method '>>='.
 --
--- /Unimplemented/
+-- >>> Stream.fold (concatMap (flip Fold.takeLE Fold.sum) (Fold.rmapM (return . fromJust) Fold.head)) $ Stream.fromList [10,9..1]
+-- 45
+--
+-- /Internal/
 --
 {-# INLINE concatMap #-}
-concatMap :: -- Monad m =>
-    (b -> Fold m a c) -> Fold m a b -> Fold m a c
-concatMap _func _fld = undefined
+concatMap :: Monad m => (b -> Fold m a c) -> Fold m a b -> Fold m a c
+concatMap f (Fold stepa initiala extracta) = Fold stepc initialc extractc
+  where
+    initialc = do
+        r <- initiala
+        case r of
+            Partial s -> return $ Partial (B s)
+            Done b -> generateInnerFold (f b)
+
+    stepc (B s) a = do
+        r <- stepa s a
+        case r of
+            Partial s' -> return $ Partial (B s')
+            Done b    -> generateInnerFold (f b)
+
+    stepc (C stepInner s extractInner) a = do
+        r <- stepInner s a
+        case r of
+            Partial sc -> return $ Partial (C stepInner sc extractInner)
+            Done c -> return $ Done c
+
+    extractc (B s) = do
+        r <- extracta s
+        initExtract (f r)
+    extractc (C _ sInner extractInner) = extractInner sInner
+
+    generateInnerFold (Fold step i e) = do
+        r <- i
+        case r of
+            Partial s -> return $ Partial (C step s e)
+            Done c -> return $ Done c
+
+    initExtract (Fold _ i e) = do
+        r <- i
+        case r of
+            Partial s -> e s
+            Done c -> return c
+
+instance Monad m => Monad (Fold m a) where
+    {-# INLINE return #-}
+    return = pure
+
+    -- | __Notice:__ In the next major release the behaviour of @<*>@ for folds
+    -- will be updated from 'teeWith' to 'splitWith' to reflect it's similarity
+    -- with parsers.
```

Comment:
Is this notice required here?

Reply:
Copy paste error from the Applicative instance